python/ray/serve/_private/replica.py

```diff
@@ -303,16 +308,20 @@ async def is_allocated(self) -> str:
                 get_component_logger_file_path(),
             )
 
-        async def is_initialized(
+        async def initialized_and_get_metadata(
             self,
             deployment_config: DeploymentConfig = None,
             _after: Optional[Any] = None,
-        ):
+        ) -> Tuple[DeploymentConfig, DeploymentVersion]:
             # Unused `_after` argument is for scheduling: passing an ObjectRef
             # allows delaying reconfiguration until after this call has returned.
             try:
-                await self._initialize_replica()
-                metadata = await self.reconfigure(deployment_config)
+                async with self._replica_init_lock:
+                    if not self._initialized:
+                        await self._initialize_replica()
+                    if deployment_config:
+                        await self.reconfigure(deployment_config)
```

Comment:
it's a behavior change to guard `reconfigure` with `if deployment_config`, please remove that (unless it's part of the fix?)

Reply:
In the recover step, we are not passing `deployment_config`.  Previously we are calling `_get_metadata`, unifying replica `recover` & `start` calling the same function, that is why we have the if check here.

Reply:
Got it, thanks for the explanation